lib/SIL/Utils/OwnershipUtils.cpp

```diff
@@ -878,6 +845,160 @@ OwnedValueIntroducer swift::getSingleOwnedValueIntroducer(SILValue inputValue) {
   llvm_unreachable("Should never hit this");
 }
 
+// Find all use points of \p guaranteedValue within its borrow scope where
+// \p guaranteedValue is not itself a borrow introducer.
+// There is no need to consider reborrows in this case.
+//
+// Accumulate results in \p usePoints, ignoring existing elements.
+//
+// Skip over nested borrow scopes. Their scope-ending instructions are their use
+// points. Transitively find all scope-ending instructions by looking through
+// nested reborrows. Nested reborrows are not use points and \p visitReborrow is
+// not called for them.
+static bool
+findInnerTransitiveGuaranteedUses(SILValue guaranteedValue,
+                                  SmallVectorImpl<Operand *> &usePoints) {
+  // Push the value's immediate uses.
+  unsigned firstOffset = usePoints.size();
+  for (Operand *use : guaranteedValue->getUses()) {
+    if (use->getOperandOwnership() != OperandOwnership::NonUse)
+      usePoints.push_back(use);
+  }
+
+  // --- Transitively follow forwarded uses and look for escapes.
+
+  // FIXME: Remove this SmallPtrSet when destructures are reborrows. Currently
+  // it forwards multiple results! This means that usePoints could grow
+  // exponentially without a membership check. It's fine to do this membership
+  // check locally in this function (within a borrow scope). It isn't needed for
+  // the immediate uses, only the transitive uses.
+  SmallPtrSet<Operand *, 16> visitedUses;
+  auto pushUse = [&](Operand *use) {
+    if (use->getOperandOwnership() != OperandOwnership::NonUse) {
+      if (visitedUses.insert(use).second)
+        usePoints.push_back(use);
+    }
+  };
+
+  // usePoints grows in this loop.
+  for (unsigned i = firstOffset; i < usePoints.size(); ++i) {
+    Operand *use = usePoints[i];
+    switch (use->getOperandOwnership()) {
+    case OperandOwnership::NonUse:
+    case OperandOwnership::TrivialUse:
+    case OperandOwnership::ForwardingConsume:
+    case OperandOwnership::DestroyingConsume:
+    case OperandOwnership::Reborrow:
+      llvm_unreachable("this operand cannot handle an inner guaranteed use");
+
+    case OperandOwnership::ForwardingUnowned:
+    case OperandOwnership::PointerEscape:
+      return false;
+
+    case OperandOwnership::InstantaneousUse:
+    case OperandOwnership::UnownedInstantaneousUse:
+    case OperandOwnership::BitwiseEscape:
+    // This Reborrow is actually the end of a nested borrow scope so handle it
+    // like a normal use of this guaranteed value.
+    case OperandOwnership::EndBorrow:
+      pushUse(use);
```

Comment:
Why am I pushing things into usePoints that are already in usePoints? Shouldn't end_borrow be in the unreachable section and we continue with the other 3 cases?